compiler/src/iree/compiler/Codegen/Common/SafeLoopInvariantCodeMotion.cpp

```diff
@@ -0,0 +1,70 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Common/Passes.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Interfaces/LoopLikeInterface.h"
+#include "mlir/Interfaces/SideEffectInterfaces.h"
+#include "mlir/Interfaces/ValueBoundsOpInterface.h"
+#include "mlir/Transforms/LoopInvariantCodeMotionUtils.h"
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_SAFELOOPINVARIANTCODEMOTIONPASS
+#include "iree/compiler/Codegen/Common/Passes.h.inc"
+
+namespace {
+/// Safer IREE loop invariant code motion (LICM) pass.
+struct SafeLoopInvariantCodeMotionPass
+    : public impl::SafeLoopInvariantCodeMotionPassBase<
+          SafeLoopInvariantCodeMotionPass> {
+  void runOnOperation() override;
+};
+} // namespace
+
+void SafeLoopInvariantCodeMotionPass::runOnOperation() {
+  // Walk through all loops in a function in innermost-loop-first order. This
+  // way, we first LICM from the inner loop, and place the ops in
+  // the outer loop, which in turn can be further LICM'ed.
+  //
+  // Hoisting is only performed on loops with guaranteed non-zero trip counts.
+  // `scf.forall` ops with mapping attributes can never be proven to have a
+  // non-zero trip count until the loop is resolved and is blanket included
+  // here.
+  getOperation()->walk([&](LoopLikeOpInterface loopLike) {
+    if (auto forallOp = dyn_cast<scf::ForallOp>(*loopLike)) {
+      if (forallOp.getMapping()) {
+        return;
+      }
+    }
+
+    // Skip loops without lower/upper bounds. There is no generic way to verify
+    // whether a loop has at least one trip so new loop types of interest can be
+    // added as needed. For example, `scf.while` needs non-trivial analysis of
+    // its condition region to know that it has at least one trip.
+    std::optional<SmallVector<OpFoldResult>> maybeLowerBounds =
+        loopLike.getLoopLowerBounds();
+    std::optional<SmallVector<OpFoldResult>> maybeUpperBounds =
+        loopLike.getLoopUpperBounds();
+    if (!maybeLowerBounds || !maybeUpperBounds) {
+      return;
+    }
+
+    // If any lower + upper bound pair cannot be definitely verified as lb < ub
+    // then the loop may have a zero trip count.
+    for (auto [lb, ub] :
+         llvm::zip_equal(*maybeLowerBounds, *maybeUpperBounds)) {
+      if (!ValueBoundsConstraintSet::compare(lb, ValueBoundsConstraintSet::LT,
+                                             ub)) {
+        return;
+      }
+    }
+
+    moveLoopInvariantCode(loopLike);
```

Comment:
Are you sure that's the case? e.g. operations with print side effects could happen once rather than multiple times in the loop.

Reply:
I haven't studied the details of the speculatability interface though and could be wrong.

Reply:
> Are you sure that's the case? e.g. operations with print side effects could happen once rather than multiple times in the loop.

printing is a memory sideeffect. speculation is for undefined behavior and infinite loops. Although reading on speculation more, a goto does not have memory effects but is not speculatable and should not be hoisted out. 

I'm probably wrong about my earlier comment and you can ignore it :)

Reply:
There's a goto op?!

Reply:
I was reading from here: https://mlir.llvm.org/docs/Rationale/SideEffectsAndSpeculation/#modeling

But reading more, actually speculation isn't supposed to handle something like goto/longjmp. Anyways, this was just a passing by comment, I can look more into it and change it later.

Reply:
Ok I see, yeah changing it later or maybe adding another case to the interface if we have cases that need it.